cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go

```diff
@@ -101,6 +103,26 @@ func setMasterTolerationOnPodSpec(podSpec *v1.PodSpec) {
 	podSpec.Tolerations = append(podSpec.Tolerations, kubeadmconstants.MasterToleration)
 }
 
+// setHostIPOnPodSpec sets the environment variable HOST_IP using downward API
+func setHostIPOnPodSpec(podSpec *v1.PodSpec) {
+	envVar := v1.EnvVar{
+		Name: "HOST_IP",
+		ValueFrom: &v1.EnvVarSource{
+			FieldRef: &v1.ObjectFieldSelector{
+				FieldPath: "status.hostIP",
+			},
+		},
+	}
+
+	podSpec.Containers[0].Env = append(podSpec.Containers[0].Env, envVar)
+
+	for i := range podSpec.Containers[0].Command {
+		if strings.Contains(podSpec.Containers[0].Command[i], "advertise-address") {
+			podSpec.Containers[0].Command[i] = "--advertise-address=$(HOST_IP)"
+		}
+	}
+}
```

Comment:
@andrewsykim in the static pod manifest for `kube-apiserver`, the advertise address is used for a liveness probe. e.g.
```yaml
livenessProbe:
      failureThreshold: 8
      httpGet:
        host: {AdvertiseAddress}
        path: /healthz
        port: 6443
        scheme: HTTPS
      initialDelaySeconds: 15
      timeoutSeconds: 15
```
If this is preserved in the selfhosted daemonset (I'm not 100% sure), we should mutate this as well.

Reply:
Good catch! Let me look into this!

Reply:
The advertise address passed in is preserved when using self hosted which is now the wrong behaviour since it should match whatever is set via downward API. Anyone know if it's possible to use downward API in a liveness probe? Alternatively we can let it default to localhost (which is the current behaviour if you do not set advertise address).

Reply:
IMO we can use 127.0.0.1 (I guess it is not possible to use downward API in a liveness probe)
@luxas @xiangpengzhao WDYT?

Reply:
127.0.0.1 SGTM. In kubemark, it also uses 127.0.0.1 for api-server probe. So I think it's ok to use it here as well.

https://github.com/kubernetes/kubernetes/blob/de2aec31d9272d18cfef9d38ad1ae42160e2faa5/test/kubemark/resources/manifests/kube-apiserver.yaml#L21-L28

OTOH, we can leave this field empty. If so, it will default to `PodIP`. And yeah, as was discussed in https://github.com/kubernetes/kubernetes/pull/56084#discussion_r152388796,  `PodIP == HostIP` in our case.

https://github.com/kubernetes/kubernetes/blob/de2aec31d9272d18cfef9d38ad1ae42160e2faa5/pkg/kubelet/prober/prober.go#L156-L159